src/IB/IBFEMethod.cpp

```diff
@@ -928,10 +928,28 @@ IBFEMethod::spreadForce(const int f_data_idx,
     }
 
     {
-        auto hierarchy = d_use_scratch_hierarchy ? d_scratch_hierarchy : d_hierarchy;
+        // If we have multiple IBMethod objects we may end up with a wider
+        // ghost region than the one required by this class. Hence, set the
+        // ghost width by just picking whatever the data actually has at the
+        // moment.
+        //
+        // f_var->getPatchDataFactory()->getGhostCellWidth() returns
+        // the wrong values - probably a SAMRAI bug. Do it the long way:
+        const Pointer<PatchLevel<NDIM> > level = hierarchy->getPatchLevel(ln);
+        IntVector<NDIM> gcw;
+        for (PatchLevel<NDIM>::Iterator p(level); p; p++)
+        {
+            const Pointer<Patch<NDIM> > patch = level->getPatch(p());
+            Pointer<PatchData<NDIM> > f_data = patch->getPatchData(f_scratch_data_idx);
+            gcw = f_data->getGhostCellWidth();
+            break;
+        }
+        // Now cover the case that this processor has no patches:
+        SAMRAI_MPI::maxReduction(gcw, NDIM);
```

Comment:
If this corresponds to a `Variable` registered with the `VariableDatabase`, I think you can ask the `VariableDatabase` for the ghost cell width directly and avoid this collective.

Reply:
Do you know how to do this? The `VariableDatabase` lets you register things with a ghost width but you cannot get them out. If you ask the `PatchDataFactory` then you get the wrong answer - I am pretty sure that this is the only way to get this value out.

Reply:
It is convoluted, but you need to access the `PatchDescriptor`, which knows about ghost cell widths.

Reply:
The `PatchDescriptor` has the `PatchDataFactory` that is actually used to construct `PatchData`. This factory should tell you the ghost cell width.

Reply:
I tried that and it returns a vector of all zeros - I think its a SAMRAI bug.

Reply:
I am sure that we use this elsewhere in the library to get ghost cell widths...

Reply:
(Also, some patch data will indeed have zero GCW.)

Reply:
Oh, yeah, if you have a patch ID and a `Patch` you should be able to get the `PatchDataFactory` directly from the patch without going through `VariableDatabase` at all. You might need to properly cast to the concrete `PatchDataFactory` class.

Reply:
The return type of `Variable<NDIM>::getPatchDataFactory()` is `Pointer<PatchDataFactory<NDIM> >` so no extra cast is required. For the following code

```cpp
        const Pointer<PatchLevel<NDIM> > level = hierarchy->getPatchLevel(ln);
        IntVector<NDIM> gcw;
        pout << "GCW from patch data factory: "
             << f_var->getPatchDataFactory()->getGhostCellWidth()
             << '\n';
        for (PatchLevel<NDIM>::Iterator p(level); p; p++)
        {
            const Pointer<Patch<NDIM> > patch = level->getPatch(p());
            Pointer<PatchData<NDIM> > f_data = patch->getPatchData(f_scratch_data_idx);
            gcw = f_data->getGhostCellWidth();
            break;
        }
        // Now cover the case that this processor has no patches:
        SAMRAI_MPI::maxReduction(gcw, NDIM);
        pout << "GCW from max reduction: "
             << gcw
             << '\n';
```
I get the output

```
GCW from patch data factory: (0,0)
GCW from max reduction: (3,3)
```

so using the patch data factory isn't going to work, even in serial.

Reply:
You cannot use the patch data factory from the variable --- it does not know what the ghost cell width is. (In SAMRAI, ghost cell widths are not properties of `Variables`.) Instead, you have to use the one associated with the `PatchDescriptor`. It _must_ know what the ghost cell width is --- otherwise it would not allocate patch data with the correct ghost cell width.

Reply:
Oh, I see. Let me try that instead.

Reply:
This is strange inconsistency in SAMRAI - `VariableDatabase::registerVariableAndContext()` takes a ghost width. Therefore it should be able to tell me the ghost width of a variable.

Nevertheless getting the patch data factory out of the patch descriptor works so lets go with that.

Reply:
The way that SAMRAI works, the `Variable` + `VariableContext` together are assigned a ghost cell width. The `Variable` by itself does not correspond to _any_ particular data storage.

Reply:
That explains it!